drivers/gpio/gpio_davinci.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2023 BeagleBoard.org Foundation
+ * Copyright (C) 2023 S Prashanth
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_davinci_gpio
+
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/arch/common/sys_bitops.h>
+#include <zephyr/devicetree.h>
+
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+/* Helper Macros for GPIO */
+#define DEV_GPIO_CFG(dev)	\
+		((const struct gpio_davinci_config * const)((dev)->config))
+#define DEV_GPIO_CFG_BASE(dev) \
+		((struct gpio_davinci_regs *)(DEV_GPIO_CFG(dev)->port_base))
+
+#define GPIO_DAVINCI_DIR_RESET_VAL	(0xFFFFFFFF)
+
+struct gpio_davinci_regs {
+	uint32_t dir;
+	uint32_t out_data;
+	uint32_t set_data;
+	uint32_t clr_data;
+	uint32_t in_data;
+	uint32_t set_ris_trig;
+	uint32_t clr_ris_tirg;
+	uint32_t set_fal_trig;
+	uint32_t clr_fal_trig;
+	uint32_t intstat;
+};
+
+struct gpio_davinci_data {
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+struct gpio_davinci_config {
+	void (*bank_config)(const struct device *dev);
+	struct gpio_driver_config common;
+	uint32_t port_base;
+	uint32_t port_num;
+};
+
+static int gpio_davinci_configure(const struct device *dev, gpio_pin_t pin,
+					gpio_flags_t flags)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_OUTPUT) != 0) {
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			regs->set_data = BIT(pin);
+		} else {
+			regs->clr_data = BIT(pin);
+		}
+
+		regs->dir &= (~(BIT(pin)));
+
+	} else {
+		regs->dir |= (BIT(pin));
+	}
+
+	return 0;
+}
+
+static int gpio_davinci_port_get_raw(const struct device *dev,
+					gpio_port_value_t *value)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	*value = regs->in_data;
+
+	return 0;
+}
+
+static int gpio_davinci_port_set_masked_raw(const struct device *dev,
```

Comment:
Each of the functions in this driver could use some documentation, for example
https://elixir.bootlin.com/linux/v6.5-rc4/source/drivers/gpio/gpio-davinci.c#L130

See how linux documents most of these gpio funcs explaining briefly what they're doing.

Reply:
Discussed in discord https://discord.com/channels/1108795636956024986/1109167658131275886/1141376293850595378